src/problems/vrp.h

```diff
@@ -10,17 +10,197 @@ All rights reserved (see LICENSE).
 
 */
 
+#include <mutex>
+#include <set>
+#include <thread>
+
+#include "algorithms/heuristics/heuristics.h"
+#include "algorithms/local_search/local_search.h"
+#include "structures/vroom/input/input.h"
 #include "structures/vroom/solution/solution.h"
 
 namespace vroom {
 
-class Input;
-
 class VRP {
   // Abstract class describing a VRP (vehicle routing problem).
 protected:
   const Input& _input;
 
+  template <class TRoute, class TSolution, class TLocalSearch>
```

Comment:
Pure nitpicking: in other template functions we use `Route` as template type to stand for either `RawRoute` or `TWRoute` so I'd be in favor of simply removing the `T` in front of template types here.

Reply:
I added the `T` prefix (which I also dislike) just to avoid clash between the `Solution` template type and the `Solution` function return type, and for consistency applied it to `Route` and `LocalSearch` as well.

I can
- rename the `TSolution` template parameter to e.g. `CandidateSolution` and drop the `T` prefixes, or
- remove the `TSolution` template parameter and replace its use throughout the function with `std::vector<Route>`

What do you think?

Reply:
Right, I did not think about the `Solution` collision. I like your second option since it removes a template type that can be trivially deduced from the other one. We can then also get rid of the `RawSolution` and `TWSolution` aliases in `cvrp.cpp` and `vrptw.cpp`. Another benefit is that it's safer to some degree as it prevents from using `solve` with inconsistent route and solution types.